### PR TITLE
Add PHPStan

### PR DESCRIPTION
**Pull request by jdrieghe**
### Added
- `phpstan/phpstan` is now part of the dependencies and CI pipeline and is enforced at level 2

### Fixed
- Incorrect types indicated by PHPStan

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -30,7 +30,8 @@
       "squizlabs/php_codesniffer": "^3.5.8",
       "phing/phing": "~2.10",
       "satooshi/php-coveralls": "~0.7",
-      "phpunit/phpunit": "^7.5"
+      "phpunit/phpunit": "^7.5",
+        "phpstan/phpstan": "^0.12.68"
```
Comment (lucwollants) [inline composer.json]: Some extra indentation.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -33,3 +33,23 @@ jobs:
 
             - name: Run Phing
               run: composer run-script phing
+
+    phpstan:
```
Comment (lucwollants) [inline .github/workflows/ci.yml:49]: I think it's worth making PHPStan part of `phing`. Because we are getting used to just run `composer phing` to execute all quality checks.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -33,3 +33,23 @@ jobs:
 
             - name: Run Phing
               run: composer run-script phing
```
Comment (lucwollants) [inline .github/workflows/ci.yml]: Can become `composer phing` 😉

File: .github/workflows/ci.yml (hunk 2 above)
Comment (jdrieghe) [reply .github/workflows/ci.yml:49]: I'm not sure I see the value of having `phing` to run these checks.
If it's simply a matter of running "all checks at once", we can solve that with a simple composer script or a makefile.

The downside of keeping phing is that it's yet another tool in the pipeline that we need to upgrade and maintain.
I'm wondering how valuable it is to keep it.

Comment (bertramakers) [reply .github/workflows/ci.yml:49]: We could also just stop using Phing of course. I'd be fine with that. 

One way or another, I'd just like to be able to run all the ci scripts together (at least locally).

Composer scripts can achieve the same goal, with one less (dev) dependency to keep up-to-date.
See https://blog.martinhujer.cz/have-you-tried-composer-scripts/

But you're right that if we use Phing, I'd add it to the `build.xml`

If we do drop Phing here, I'd also make a ticket to remove it in other repo's later. In the meantime we can remove it on an individual basis when we work on a repo of course.

Comment (bertramakers) [reply .github/workflows/ci.yml:49]: Ah I was typing my comment while you posted yours @jdrieghe 😄

Comment (bertramakers) [reply .github/workflows/ci.yml:49]: I can also see the value of running them separately on Github Actions. I'm only wondering if we also have a `ci` script like in the article above, for running everything locally, if those two lists of checks wouldn't get out of sync easily?

Comment (jdrieghe) [reply .github/workflows/ci.yml:49]: Additionally, I don't think we need to run PHP CS and phpstan against all of the supported PHP versions.

Comment (bertramakers) [reply .github/workflows/ci.yml:49]: True. So we would run the separately on Github Actions (and not always all of them for every PHP version), but also have a single script that combines them locally? I guess we can try that and deal with inconsistencies of local vs Github Actions if they pop up and if need be change strategy then.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -33,3 +33,23 @@ jobs:
...
+              uses: actions/checkout@v2
+
+            - name: 🐘 Install PHP
+              uses: shivammathur/setup-php@v2
+              with:
+                  php-version: '7.4'
+                  tools: composer:v2
+
+            - name: 📦 Install dependencies
+              if: steps.composer-cache.outputs.cache-hit != 'true'
```
Comment (bertramakers) [inline .github/workflows/ci.yml]: I think this `if` can be removed, also from the installation of dependencies in the other job. I removed the caching in a previous PR because this library has no `.lock` file so the cache never got invalidated. And we can't cache based on `composer.json`, because then it wouldn't install the latest versions (matching the constraints of the versions in composer.json) because it would have cached older versions and not invalidate them because `composer.json` didn't change, but there _are_ newer versions available which would be installed in other circumstances like when someone adds this package to a project.

Long story short: I don't think we should cache composer dependencies if we don't have a `.lock` file.

And the reason we don't have a `.lock` file in our libraries is so we don't get a false sense of security that everything works, because when you add this package to a project the `.lock` file in it will be ignored and a newer version of a dependency might be installed.

File: composer.json [hunk 5]
```diff
@@ -35,7 +35,11 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "scripts": {
-      "phing": "./vendor/bin/phing test",
+      "pre-commit": [
```
Comment (bertramakers) [inline composer.json]: The idea is that we run this as `composer pre-commit` manually, or does it always run for every commit you do?
I'd prefer it not to run automatically for every commit because sometimes commits are just temporarily broken by design (e.g. TDD, or when refactoring but splitting up the work in multiple commits)

Comment (bertramakers) [reply composer.json]: And if we run it manually it might be nice to have something shorter like `composer ci`?

Comment (jdrieghe) [reply composer.json]: I couldn't really find a good name but I like ci. It's short and covers the load.

**Review by bertramakers — APPROVED** [review]
Much more elegant, I'm looking forward to also having this in the other repos! 👏
